crypto/provider_core.c

```diff
@@ -259,6 +269,79 @@ int ossl_provider_disable_fallback_loading(OSSL_LIB_CTX *libctx)
     return 0;
 }
 
+#define BUILTINS_BLOCK_SIZE     10
+
+int ossl_provider_info_add_to_store(OSSL_LIB_CTX *libctx,
+                                    OSSL_PROVIDER_INFO *entry)
+{
+    struct provider_store_st *store = get_provider_store(libctx);
+    int ret = 0;
+
+    if (entry->name == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    if (store == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+
+    if (!CRYPTO_THREAD_write_lock(store->lock))
+        return 0;
+    if (store->provinfosz == 0) {
+        store->provinfo = OPENSSL_zalloc(sizeof(*store->provinfo)
+                                         * BUILTINS_BLOCK_SIZE);
+        if (store->provinfo == NULL) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE);
+            goto err;
+        }
+        store->provinfosz = BUILTINS_BLOCK_SIZE;
+    } else if (store->numprovinfo == store->provinfosz) {
+        OSSL_PROVIDER_INFO *tmpbuiltins;
+        size_t newsz = store->provinfosz + BUILTINS_BLOCK_SIZE;
+
+        tmpbuiltins = OPENSSL_realloc(store->provinfo,
+                                      sizeof(*store->provinfo) * newsz);
+        if (tmpbuiltins == NULL) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE);
+            goto err;
+        }
+        store->provinfo = tmpbuiltins;
+        store->provinfosz = newsz;
+    }
+    store->provinfo[store->numprovinfo] = *entry;
+    store->numprovinfo++;
+
+    ret = 1;
+ err:
+    CRYPTO_THREAD_unlock(store->lock);
+    return ret;
+}
+
+int OSSL_PROVIDER_add_builtin(OSSL_LIB_CTX *libctx, const char *name,
+                              OSSL_provider_init_fn *init_fn)
```

Comment:
Bleah.  The intention was that `provider.c` would have all the public facing provider functions, while `provider_core.c` would contain internal functions.  Apparently, that "layering" didn't catch, and the point of having the two separate source files is becoming lost.

(and if we decide to merge the two files, it will be yet another huge source file...  we never learn 😕 )

Reply:
This is the first time I heard of such a distinction.  If we aren't aware of something, we can't be expected to adhere to it....

Reply:
I'm kinda amazed that the mere presence of two files doesn't make you wonder why...  if you look at their history, it should be quite visible that one has all public functions and the other not.  Unfortunately, this started sliding some time ago already; I didn't notice at the time.

Reply:
Its news to me too. Actually though for this particular function I've moved it back again. Earlier commits in this PR needed to move it because the function needed access to internals of the provider structure. However the final version of the function no longer needs this. The same doesn't apply to `OSSL_PROVIDER_available` though - which still needs to move. An alternative would be to just have an internal function that does the same job as `OSSL_PROVIDER_available`, and then the public function could be a very simple thin wrapper that is the only caller of the internal function. However this seems a bit silly to me.

Reply:
> ... and then the public function could be a very simple thin wrapper that is the only caller of the internal function.

That's what was done for a number of them, so it wouldn't be strange. 

> However this seems a bit silly to me

And that basically tells me that the separation that I intended is just me. Fine. Then we should eventually merge provider.c and provider_core.c
(no,not in this pr)